libfwupd/fwupd-hsi-attr.c

```diff
@@ -0,0 +1,730 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+//#include <glib-object.h>
+#include <gio/gio.h>
+//#include <string.h>
+
+#include "fwupd-common-private.h"
+#include "fwupd-enums-private.h"
+//#include "fwupd-error.h"
+#include "fwupd-hsi-attr-private.h"
+
+/**
+ * SECTION:fwupd-hsi-attr
+ *
+ * An object that represents an Host Security ID attribute.
+ */
+
+static void fwupd_hsi_attr_finalize	 (GObject *object);
+
+typedef struct {
+	gchar				*appstream_id;
+	GPtrArray			*obsoletes;
+	gchar				*name;
+	gchar				*summary;
+	gchar				*uri;
+	guint32				 number;
+	FwupdHsiAttrFlags		 flags;
+} FwupdHsiAttrPrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (FwupdHsiAttr, fwupd_hsi_attr, G_TYPE_OBJECT)
+#define GET_PRIVATE(o) (fwupd_hsi_attr_get_instance_private (o))
+
+static const gchar *
+fwupd_hsi_attr_flag_to_string (FwupdHsiAttrFlags flag)
+{
+	if (flag == FWUPD_HSI_ATTR_FLAG_NONE)
+		return "none";
+	if (flag == FWUPD_HSI_ATTR_FLAG_SUCCESS)
+		return "success";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_UPDATES)
+		return "runtime-updates";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_ATTESTATION)
+		return "runtime-attestation";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_ISSUE)
+		return "runtime-issue";
+	if (flag == FWUPD_HSI_ATTR_FLAG_RUNTIME_UNTRUSTED)
+		return "runtime-untrusted";
+	return NULL;
+}
+
+/**
+ * fwupd_hsi_attr_get_obsoletes:
+ * @self: A #FwupdHsiAttr
+ *
+ * Gets the self obsoletes.
+ *
+ * Returns: (element-type utf8) (transfer none): the obsoletes, which may be empty
+ *
+ * Since: 1.4.2
+ **/
+GPtrArray *
+fwupd_hsi_attr_get_obsoletes (FwupdHsiAttr *self)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_val_if_fail (FWUPD_IS_HSI_ATTR (self), NULL);
+	return priv->obsoletes;
+}
+
+/**
+ * fwupd_hsi_attr_add_obsolete:
+ * @self: A #FwupdHsiAttr
+ * @appstream_id: the attribute appstream_id
+ *
+ * Sets the attribute appstream_id.
+ *
+ * Since: 1.4.2
+ **/
+void
+fwupd_hsi_attr_add_obsolete (FwupdHsiAttr *self, const gchar *appstream_id)
+{
+	FwupdHsiAttrPrivate *priv = GET_PRIVATE (self);
+	g_return_if_fail (FWUPD_IS_HSI_ATTR (self));
+	g_return_if_fail (appstream_id != NULL);
+	for (guint i = 0; i < priv->obsoletes->len; i++) {
+		const gchar *obsolete_tmp = g_ptr_array_index (priv->obsoletes, i);
+		if (g_strcmp0 (obsolete_tmp, appstream_id) == 0)
+			return;
+	}
```

Comment:
seems like this should just be `fwupd_hsi_attr_has_obsolete` rather than code duplication

Reply:
Agreed, fixed.